Ruby refactoring

MrGoumX opened the pull request:
Hello! During #26, I saw several improvements that can be made (imo) in the codebase, so here is the PR

PS: There are several things that you might not like, If there is something that you don't like we can remove it.

smortex reviewed (commented):
Regarding style, it would be really cool to configure rubocop to enforce it as part of CI to have a consistent style on the long run.

I added pointers to relevant docs, but we can probably use the config used in our modules as a sane base for this code:
https://github.com/voxpupuli/modulesync_config/blob/master/moduleroot/.rubocop.yml.erb

One commit per file does not make it easy to follow the changes :sob: Please consider commiing each change (or group of related changes) as single commits with proper commit message to make it easier to review :smile:

ekohl commented on lib/puppet_metadata/aio.rb:
```diff
@@ -1,21 +1,21 @@
-  class AIO 
+  class AIO
     COMPATIBLE = {
-      'AlmaLinux'   => 'RedHat',
```
I'd be fine with the `key` style.

MrGoumX replied:
Restored to original.

ekohl commented on lib/puppet_metadata/aio.rb:
```diff
@@ -1,21 +1,21 @@
         '6' => 5..7,
         '7' => 5..7,
-        '8' => 5..7,
+        '8' => 5..7
```
I'm a strong proponent of the `comma` style (so the current code).

MrGoumX replied:
Restored to original and followed this style on some new hashes.

ekohl commented on spec/metadata_spec.rb:
```diff
@@ -65,15 +65,15 @@
             {
               operatingsystem: 'CentOS',
-              operatingsystemrelease: ['7', '8'],
+              operatingsystemrelease: %w[7 8],
```
No strong preference from me here.

MrGoumX replied:
Restored to original.

MrGoumX commented:
I have changed the order of commits and their names. The changes are:

* Freeze hashes
* Updated the implementation of Beaker and GithubActions
* Use class << self in class methods

ekohl commented on lib/puppet_metadata/beaker.rb:
```diff
@@ -32,34 +44,29 @@ def self.os_release_to_setfile(os, release, use_fqdn: false, pidfile_workaround:
+      return '' unless options.any?
+
+      "{#{options.map { |key, value| "#{key}=#{value}" }.join(',')}}"
+    end
```
Wouldn't it be better to have `build_setfile(name, options)`?

MrGoumX replied:
I updated the implementation. If you still don't like it I'm going to restore it. 

Also I found the comment I deleted, that was my mistake. If you want to me to document any changes just tell me to add it 😛

ekohl commented on lib/puppet_metadata/github_actions.rb:
```diff
@@ -1,5 +1,12 @@
 module PuppetMetadata
   class GithubActions
+    PUPPET_RUBY_VERSIONS = {
```
I like making this a constant hash, but it's not really a Github Actions thing. It's really an `AIO` thing. It should be defined in `PuppetMetadata::AIO`.

MrGoumX replied:
It has been moved.

MrGoumX commented on lib/puppet_metadata/github_actions.rb:
```diff
@@ -52,32 +59,16 @@ def puppet_unit_test_matrix
-      matrix_include = []
-
-      metadata.operatingsystems.each do |os, releases|
+      metadata.operatingsystems.each_with_object([]) do |(os, releases), arr|
```
I changed the name back to `matrix_include`. If you don't like it we can restore it to the original.

MrGoumX commented on lib/puppet_metadata/beaker.rb:
```diff
@@ -32,34 +44,29 @@ def self.os_release_to_setfile(os, release, use_fqdn: false, pidfile_workaround:
+        return if IMAGES_EXCEPTIONS[os]&.include?(release)
 
-      human_name = "#{os} #{release}"
+        options[:image] = IMAGES.dig(os, release) unless IMAGES.dig(os, release).nil?
```
I agree this is better.

MrGoumX commented on lib/puppet_metadata/beaker.rb:
```diff
@@ -32,34 +44,29 @@ def self.os_release_to_setfile(os, release, use_fqdn: false, pidfile_workaround:
+      if pidfile_workaround?(pidfile_workaround, os)
+        return if IMAGES_EXCEPTIONS[os]&.include?(release)
 
-      human_name = "#{os} #{release}"
```
Restored

smortex commented:
Thanks a lot @MrGoumX for splitting the commits by what they do :+1: 

I fail to spot the change of the static methods (`def self.foo`) to the blocks with what I identify as an "eigenclass" `class << self; def foo`. I remember vaguely using similar things while scratching my head in meta-programing scenarios but here I fail to spot what is the goal you want to achieve?

Apart from this point I don't understand, :100: !

ekohl commented on lib/puppet_metadata/beaker.rb:
```diff
@@ -4,62 +4,77 @@ module PuppetMetadata
-    # @return [Array<(String, String)>] The beaker setfile description with a readable name
-    def self.os_release_to_setfile(os, release, use_fqdn: false, pidfile_workaround: false)
-      return unless os_supported?(os)
+    IMAGES = {
```
Perhaps we can add a comment here that these images have an older systemd which isn't strict and that's why they do work with `PIDFile`. Previously I considered it an implementation detail but now that they're constants it should be documented. Thinking more about it, perhaps we can also find a better name than `IMAGES`. The most literal one would be `IMAGES_WITH_PIDFILE_COMPATIBLE_SYSTEMD` but perhaps there's a better name. Something like `PIDFILE_COMPATIBLE_IMAGES`

MrGoumX replied:
Also changed the variable name and added a comment that hopefully adds some context to the variable.

ekohl commented on lib/puppet_metadata/beaker.rb:
```diff
@@ -4,62 +4,77 @@ module PuppetMetadata
-      name = "#{os.downcase}#{release.tr('.', '')}-64"
+    # There is no CentOS 8 image that works with PIDFile in systemd
+    # unit files
+    IMAGES_EXCEPTIONS = {
```
Here we can also make the name more explicit. How about `PIDFILE_INCOMPATIBLE_OS_VERSIONS` or just `PIDFILE_INCOMPATIBLE`.

MrGoumX replied:
Changed the variable name!

ekohl reviewed (commented):
Overall I think I'm happy with the changes except a bit of bikeshedding. Let me know what you think.

MrGoumX commented:
@smortex The only reason I changed these class methods to use `class << self` is to not write `self.method` every time. It's just a personal preference.

ekohl reviewed (approved):
Looks good. I see you already prepared for squashing so I'd be happy to merge if you did so.

MrGoumX commented:
Changes squashed!

ekohl commented:
Thanks!
